stock_reserve/model/stock_reserve.py

```diff
@@ -0,0 +1,197 @@
+              Note: `default_picking_type_id` may be present in context,
+              so code that looks for default `location_id` is implemented here,
+              because it relies on already calculated default
+              `picking_type_id`.
+        """
+        # if there is 'location_id' field requested, ensure that
+        # picking_type_id is also requested, because it is required
+        # to compute location_id
+        if ('location_id' in fields_list and
+                'picking_type_id' not in fields_list):
+            fields_list = fields_list + ['picking_type_id']
+
+        res = super(StockReservation, self).default_get(fields_list)
+
+        if 'product_qty' in res:
+            del res['product_qty']
+
+        # At this point picking_type_id and location_id
+        # should be computed in default way:
+        #     1. look up context
+        #     2. look up ir_values
+        #     3. look up property fields
+        #     4. look up field.default
+        #     5. delegate to parent model
+        #
+        # If picking_type_id is present and location_id is not, try to find
+        # default value for location_id
+        if not res.get('picking_type_id', None):
+            res['picking_type_id'] = self._default_picking_type_id()
+
+        picking_type_id = res.get('picking_type_id')
+        if picking_type_id and not res.get('location_id', False):
+            picking = self.env['stock.picking'].new(
+                {'picking_type_id': picking_type_id})
+            picking.onchange_picking_type()
+            res['location_id'] = picking.location_id.id
+        if 'location_dest_id' in fields_list:
+            res['location_dest_id'] = self._default_location_dest_id()
+        if 'product_uom_qty' in fields_list:
+            res['product_uom_qty'] = 1.0
+        return res
+
+    @api.model
+    def get_location_from_ref(self, ref):
+        """ Get a location from a xmlid if allowed
+        :param ref: tuple (module, xmlid)
+        """
+        try:
+            location = self.env.ref(ref, raise_if_not_found=True)
+            location.check_access_rule('read')
+            location_id = location.id
+        except (except_orm, ValueError):
+            location_id = False
+        return location_id
+
+    @api.model
+    def _default_picking_type_id(self):
+        ref = 'stock.picking_type_out'
+        return self.env.ref(ref, raise_if_not_found=False).id
+
+    @api.model
+    def _default_location_dest_id(self):
+        ref = 'stock_reserve.stock_location_reservation'
+        return self.get_location_from_ref(ref)
+
+    @api.multi
+    def reserve(self):
+        """ Confirm reservations
+
+        The reservation is done using the default UOM of the product.
+        A date until which the product is reserved can be specified.
+        """
+        self.write({'date_expected': fields.Datetime.now()})
+        self.mapped('move_id')._action_confirm()
```

Comment:
here you have the problem

I tested the following:
- If the product is storable, everything works.
- If the product is consumable it fails.
- If the product is storable but you have not enough qty available, it fails again.

So, you need to:
- Ad a filter in product to allow only storable products.
- Add a constrain in qty to reserve to avoid the user to ask a reservation to more products than available.
- Catch the error in order to avoid a problem is anyone else consumes the available qty before you confirm.

Reply:
When the product is storable and you have the qty to reserve,not really everything works.When I create a sale order which contain two storable products,when I click the reserve stock button to create reversions and add a validity date,it will raise the error.

Reply:
You are right, I didn't tested that as I thought the problem was the same.

The problem seems to be that you have several records and action confirm fails.

Thy doing a loop, instead of lines 140 and 141 try:

for m in self.mapped('move_id')
    m._action_confirm()
    m.picking_id.action_assign()

It's posible that it will fail because you may not have a picking on the move, if that's the case, try:

for m in self.mapped('move_id')
    m._action_confirm()
    if m.picking_id:
        m.picking_id.action_assign()

Reply:
I was thinking about this and most of the time we will have one picking for all moves ir no move at all, si, de can do:

```
for m in self.mapped('move_id')
    m._action_confirm()
for p in self.mapped('move_id.picking_id')
        p.action_assign()
```

Reply:
OK,I'll try.Thank you very much.

Reply:
```python
for m in self.mapped('move_id'):
    m._action_confirm()
for p in self.mapped('move_id.picking_id'):
     p.action_assign()
```
This didn't solve the issue that create a stock reservation from a sale order with multi lines.

Reply:
Ok, I checked the problem. The problem is not because of multiple lines.

In v12 _action_confirm is defined like this _action_confirm(self, merge=True, merge_into=False), while in v9, it def action_confirm(self).

As you can see, you have 2 new params in the definition, one of them is True by default.

For that reason, it merges the move with deletes the original move.
Try this: `m._action_confirm(merge=False, merge_into=False)`

Reply:
Great!!!It works.You are so helpful,thank you very much.